### PR TITLE
Mpi/cleanup mpi lifetime management

### PR DESCRIPTION
**Pull request by jcotela**
@pooyan-dadvand this is what we discussed in person regarding MPI initialization/finalization. This removes the need for a separate `MPIEnvironment` singleton, delegating MPI lifetime management to a member in ParallelEnvironment (initialized once a MPIDataCommunicator is created).

Note that this PR is into the mpi/experimental-split-communicator branch: it is so closely linked to the changes there that I would need to duplicate much of the work in that PR to achieve the same result. Still, I think it is worth to review it separately.

### CODE REVIEW

**Comment by philbucher** [comment]
ping @pooyan-dadvand this is the main blocker for finishing the MPI-capabilities of the CoSimApp

File: kratos/includes/parallel_environment.h [hunk 1]
```diff
@@ -32,6 +32,23 @@ namespace Kratos
...
+class KRATOS_API(KRATOS_CORE) EnvironmentManager
+{
+  public:
+    typedef std::unique_ptr<EnvironmentManager> Pointer;
+
+    EnvironmentManager(EnvironmentManager& rOther) = delete;
+
+    virtual ~EnvironmentManager() = default;
+
+    virtual bool IsInitialized() const = 0;
```
Comment (pooyan-dadvand) [inline kratos/includes/parallel_environment.h:44]: I don't understand how being an abstract class compiles in this case? It should give some error in `bool ParallelEnvironment::MPIIsInitializedDetail() const`. Isn't it?

Comment (pooyan-dadvand) [reply kratos/includes/parallel_environment.h:44]: No I see that it will check it on runtime.

Sorry for the noise!

**Review by pooyan-dadvand — APPROVED** [review]
:+1:

**Comment by jcotela** [comment]
@philbucher I'll go ahead and merge so that you are not blocked, but I haven't tested this branch in a long time... I apologize in advance in case something breaks in current master :fearful:

**Pull request merged**